source/plugin/optional/basic.rst

```diff
@@ -0,0 +1,195 @@
+all the more poingant when the API in question is a layer over an extremely unstable base product. Fortunately,
+there is a better way:
+
+2. Optional and the Explicit Nullable Contract
+==============================================
+
+As mentioned above, APIs for Minecraft are in a difficult situation. Ultimately they need to provide a platform with
+a *reasonable amount of implied stability* atop a platform (the game) with *absolutely no amount of implied stability*.
+Thus any API for Minecraft needs to be designed with full awareness that any aspect of the game is liable to change at
+any time for any reason in any way imaginable; up to and including being removed altogether!
+
+This volatility is what leads to the problem with nullable method contracts described above.
+
+`Optional` solves the above problems by replacing *implicit contracts* with *explicit* ones. The API never advertises,
+*"here is your object, kthxbai"*, instead it presents accessors with a
+*"here is a box which may or may not contain the object you asked for, ymmv"*.
+
+.. image:: /images/optionals3.png
+
+By encoding the possibility of returning ``null`` into an explicit contract, we replace the concept of
+*null checking* with the more nuanced concept of *may not exist*. We also stipulate this contract *from day one*.
+
+So what does this mean?
+
+In a nutshell, no longer do plugin authors have to worry about may/may not return ``null``, instead the very
+possibility of a particular object not being available becomes encoded in the very fabric of their plugin code. This
+has the same level of inherent safety as just null-checking the hell out of everything, but with the benefit of much
+more elegant and readable code in order to do so.
+
+To see why, let's take a look at the above example, converted to use a ``getFoo`` method which returns
+``Optional<Foo>`` instead:
+
+.. code-block:: java
+
+    public void someEventHandler(Entity someEntity) {
+        Optional<Foo> entityFoo = someEntity.getFoo();
+        if (entityFoo.isPresent()) {
+            entityFoo.get().bar();
+        }
+    }
+
+You may note that this example looks very much like a standard null-check, however the use of ``Optional`` actually
+carries a little more information in the same amount of code. For example, it is not necessary for someone reading
+the above code to check the method contract, it is clear that the method may not return a value, and the handling
+of the value's absence is explicit and clear.
+
+So what? Our explicit contract in this case results in basically the same amount of code as a null check - albeit
+one that is contractually *enforced* by the getter. *"Whoop de do,"* you say, *"so what?"*
+
+Well the `Optional` boxing allows us to take some of the traditionally more awkward aspects of null-checking and
+make them orders of magnitude more elegant: consider the following code:
```

Comment:
Not sure about this since I think it fits the style of the text. I'll leave this for someone else to decide.